test-tests/shared/src/test/scala/zio/test/TestAspectSpec.scala

```diff
@@ -0,0 +1,118 @@
+        },
+        testM("jsOnly runs tests only on ScalaJS") {
+          val spec   = test("Javascript-only")(assert(TestPlatform.isJS, isTrue)) @@ jsOnly
+          val result = if (TestPlatform.isJS) succeeded(spec) else ignored(spec)
+          assertM(result, isTrue)
+        },
+        testM("jvm applies test aspect only on jvm") {
+          for {
+            ref    <- Ref.make(false)
+            spec   = test("test")(assert(true, isTrue)) @@ jvm(after(ref.set(true)))
+            _      <- execute(spec)
+            result <- ref.get
+          } yield assert(if (TestPlatform.isJVM) result else !result, isTrue)
+        },
+        testM("jvmOnly runs tests only on the JVM") {
+          val spec   = test("JVM-only")(assert(TestPlatform.isJVM, isTrue)) @@ jvmOnly
+          val result = if (TestPlatform.isJVM) succeeded(spec) else ignored(spec)
+          assertM(result, isTrue)
+        },
+        test("failure makes a test pass if the result was a failure") {
+          assert(throw new java.lang.Exception("boom"), isFalse)
+        } @@ failure,
+        test("failure makes a test pass if it died with a specified failure") {
+          assert(throw nullPointerException, isFalse)
+        } @@ failure(diesWith(equalTo(nullPointerException))),
+        test("failure does not make a test pass if it failed with an unexpected exception") {
+          assert(throw new NullPointerException(), isFalse)
+        } @@ failure(failsWithException[IllegalArgumentException])
+          @@ failure,
+        test("failure does not make a test pass if the specified failure does not match") {
+          assert(throw new RuntimeException(), isFalse)
+        } @@ failure(
+          isCase[TestFailure[String], Cause[String]]("Runtime", {
+            case TestFailure.Runtime(e) => Some(e); case _ => None
+          }, equalTo(Cause.fail("boom")))
+        ) @@ failure,
+        test("failure makes tests pass on any assertion failure") {
+          assert(true, equalTo(false))
+        } @@ failure,
+        test("failure makes tests pass on an expected assertion failure") {
+          assert(true, equalTo(false))
+        } @@ failure(
+          isCase[TestFailure[Any], Any](
+            "Assertion",
+            { case TestFailure.Assertion(result) => Some(result); case _ => None },
+            anything
+          )
+        ),
+        testM("timeout makes tests fail after given duration") {
+          assertM(ZIO.never *> ZIO.unit, equalTo(()))
+        } @@ timeout(1.nanos)
+          @@ failure(diesWith(equalTo(timeoutFailure))),
+        testM("timeout reports problem with interruption") {
+          assertM(ZIO.never.uninterruptible *> ZIO.unit, equalTo(()))
+        } @@ timeout(10.millis, 1.nano)
+          @@ failure(diesWith(equalTo(interruptionTimeoutFailure)))
+      )
+    )
+object TestAspectSpecUtil {
+  def failsWithException[E](implicit ct: ClassTag[E]): Assertion[TestFailure[E]] =
```

Comment:
You can implement this in terms of `diesWith`. Let's update the name to something like `diesWithSubtypeOf` so it is clear how this and `diesWith` are related if someone else is working in this file.

Reply:
that makes things much neater!